### ISSUE TITLE
Class methods such as Class::->new give errors in Modules::RequireExplicitInclusion

### ISSUE THREAD

**Issue by aaronpriven** [Perl-Critic/Perl-Critic-StricterSubs#3]
Hi. Under "Invoking class methods" in "perldoc perlobj", it says one can disambiguate class names by using a trailing :: after the class name, e.g., "Class->new" is ambiguous but "Class::->new" is not.  However, P::C::P::Modules::RequireExplicitInclusion treats the :: as part of the class name, so that it says that the package "Class::" isn't found. Ideally this would be allowed.

An example with a core class method:

```
#!/usr/bin/env perl

use strict;
use warnings;

our $VERSION = 0.001;

use Net::Netrc;
$mach = Net::Netrc::->lookup('some.machine');
```

Changing Net::Netrc:: to Net::Netrc avoids the error.

Thanks for making this program. I looked at the code to try to come up with a solution but PPI is over my head :(

**Comment by petdance** [comment]
I've never seen this done before.  Is this something folks really do?  Is `Foo::Bar::` just treated as `Foo::Bar` inside of Perl?  The parser just ignores the last `::`?  /cc @sawyerx

**Comment by petdance** [comment]
I've never seen this done before.  Is this something folks really do?  Is `Foo::Bar::` just treated as `Foo::Bar` inside of Perl?  The parser just ignores the last `::`?

Does this get used anywhere else?  Can you do `use Foo::Bar:;`?  Would you want to?

**Comment by aaronpriven** [comment]
I don’t remember now where I saw this usage recommended, but I didn’t invent it. There is an ambiguity in Bar->x — does it mean call Bar the sub first, or just ?

This is going to hit few people because most people don’t have a sub Bar in the same module where they have a package Bar. But it seems like a good thing for StricterSubs to support.

—

#!/usr/bin/env perl

use feature 'say';

package Bar;

sub method {
   say "This is Bar::method";
}

package Foo;

sub Bar {
  return "Foo";
}

sub method {
  say "This is Foo::method";
}

Bar->method;
# This is Foo::method
Bar::->method;
# This is Bar::method





-- 
Aaron Priven, aaron@priven.com, www.priven.com/aaron

**Comment by aaronpriven** [comment]
Well, I don't know how many people actually do this, but it is recommended in the Perl object documentation, as I wrote originally. There is an ambiguity in Bar->x — does it mean call Bar the sub first, or just use the package Bar as the class?

This is going to hit few people because most people don’t have a sub Bar in the same module where they have a package Bar. But it seems like a good thing for StricterSubs to support.

````perl
#!/usr/bin/env perl

use feature 'say';

package Bar;

sub method {
   say "This is Bar::method";
}

package Foo;

sub Bar {
  return "Foo";
}

sub method {
  say "This is Foo::method";
}

Bar->method;
# This is Foo::method
Bar::->method;
# This is Bar::method


````

**Comment by aaronpriven** [comment]
For reference:

https://metacpan.org/pod/Perl::Critic::Policy::ValuesAndExpressions::ProhibitBarewordDoubleColon

https://metacpan.org/pod/perl5005delta#Foo::-can-be-used-as-implicitly-quoted-package-name

**Comment by xsawyerx** [comment]
Sorry. My handle is 'xsawyerx', so I didn't receive the mention.

Abigail famously calls classes with `::` at the end to clarify he means a full namespace and not a class. It then checks them during compile-time, so that's an added benefit.

```perl
$ perl -Mstrict -wc -e'Foo::->new'
Bareword "Foo::" refers to nonexistent package at -e line 1.
-e syntax OK
$ perl -Mstrict -wc -e'Foo->new'
-e syntax OK
```

**Comment by oalders** [comment]
I just ran up against this today because we use this pattern occasionally at `$work`. FWIW I did write up a blog post on this recently: https://www.olafalders.com/2022/03/30/making-dynamically-required-package-names-more-discoverable-in-perl/

I would love for the warnings I'm seeing no longer to be there. :)

> Use of uninitialized value $package in exists at /opt/perl5.34.1/lib/site_perl/5.34.1/Perl/Critic/Policy/Modules/RequireExplicitInclusion.pm line 194

**Comment by petdance** [comment]
This has been fixed in 0.06.
